src/System.Private.ServiceModel/src/System/IdentityModel/Tokens/SecurityToken.cs

```diff
@@ -14,5 +13,42 @@ public abstract class SecurityToken
         public abstract ReadOnlyCollection<SecurityKey> SecurityKeys { get; }
         public abstract DateTime ValidFrom { get; }
         public abstract DateTime ValidTo { get; }
+
+        public virtual bool CanCreateKeyIdentifierClause<T>() where T : SecurityKeyIdentifierClause
+        {
+            return ((typeof(T) == typeof(LocalIdKeyIdentifierClause)) && CanCreateLocalKeyIdentifierClause());
+        }
+
+        public virtual T CreateKeyIdentifierClause<T>() where T : SecurityKeyIdentifierClause
+        {
+            if ((typeof(T) == typeof(LocalIdKeyIdentifierClause)) && CanCreateLocalKeyIdentifierClause())
+                return new LocalIdKeyIdentifierClause(this.Id, this.GetType()) as T;
+
+            throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(
+                SR.Format(SR.TokenDoesNotSupportKeyIdentifierClauseCreation, GetType().Name, typeof(T).Name)));
+        }
+
+        public virtual bool MatchesKeyIdentifierClause(SecurityKeyIdentifierClause keyIdentifierClause)
+        {
+            LocalIdKeyIdentifierClause localKeyIdentifierClause = keyIdentifierClause as LocalIdKeyIdentifierClause;
+            if (localKeyIdentifierClause != null)
+                return localKeyIdentifierClause.Matches(this.Id, this.GetType());
+
+            return false;
+        }
+
+        public virtual SecurityKey ResolveKeyIdentifierClause(SecurityKeyIdentifierClause keyIdentifierClause)
+        {
+            if (this.SecurityKeys.Count != 0 && MatchesKeyIdentifierClause(keyIdentifierClause))
+                return this.SecurityKeys[0];
+
+            return null;
+        }
+
+        bool CanCreateLocalKeyIdentifierClause()
+        {
+            return (this.Id != null);
+        }
     }
 }
```

Comment:
If you can call the method with the generic type, then it must be able to get the Type object. I'm not seeing any dependence on the RuntimeType so I would expect it to be safe.

---
In reply to: [78184790](https://github.com/dotnet/wcf/pull/1515#discussion_r78184790) [](ancestors = 78184790,78148023)